### ISSUE TITLE
Feedback needed: checkers using Vim's embedded Perl and Python interpreters

### ISSUE THREAD

**Issue by lcd047** [scrooloose/syntastic#1351]
Branch [embedded](https://github.com/scrooloose/syntastic/tree/embedded) adds three experimental checkers: `vimperlcritic`, `vimpython`, and `vimpython3`.  These use Vim's embedded Perl and Python interpreters, and check the contents of the current buffer, rather than the corresponding file on disk.  Thus they can update the list of errors without writing the file first (which of course is only interesting in passive mode), and since they do the checks in memory, they are faster then the corresponding `perlcritic` and `python` checkers (again, this is only interesting for `perlcritic`, since `python` is pretty fast anyway).  On the down side, they need more memory than the external checkers.

In detail:

* `vimperlcritic` is an in-memory version of `perlcritic`.  It needs a version of Vim compiled with `+perl`, and of course the `Perl::Critic` module.  It has two configuration options: `g:syntastic_perl_vimperlcritic_thres` (default: 5), which is the analogue of `perlcritic`'s `g:syntastic_perl_perlcritic_thres`, and `g:syntastic_perl_vimperlcritic_format` (default: `(%s) %m (%e)`), which is the format of error messages (see `--verbose` in `perlcritic(1)`'s manual for a list of supported `%`-formats).  It doesn't support any other options, but it does try to load your `.perlcritic`, so you can still tune its output to your heart's content by editing that file.
* `vimpython` is the in-memory version of `python`.  It needs a version of Vim compiled with `+python`, and it's a faster version of `python`.  Since the `python` checker itself is only useful for detecting syntax errors, `vimpython` is pretty much just a proof of concept at this point.  You can however try running it from a `CursorHold` autocommand.
* `vimpython3` is the in-memory version of `python` with the Python 3 interpreter.  It needs a version of Vim compiled with `+python3`, and it's otherwise identical to `vimpython`.

If there is enough interest for this line of thinking, other in-memory checkers are possible.  I might also spend time designing a better user interface.  Comments welcome.

**Comment by myint** [comment]
`vimpython3` seems t work for me. The first time I run `:SyntasticInfo`, it gives the following error:

```
Syntastic version: 3.6.0-50 (Vim 704, Darwin)
Info for filetype: python
Global mode: active
Filetype python is active
The current file will be checked automatically
Error detected while processing function SyntasticInfo..83..82..20..SyntaxCheckers_python_vimpython3_IsAvailable:
line    2:
E319: Sorry, the command is not available in this version:         python import vim
Available checkers: pep257 pep8 py3kwarn pyflakes pyflakes_with_warnings pylint python vimpython3
Currently enabled checker: vimpython3
Press ENTER or type command to continue
```

Here is a fix that works for me:

```diff
diff --git a/syntax_checkers/python/vimpython3.vim b/syntax_checkers/python/vimpython3.vim
index edc6a9e..f514a03 100644
--- a/syntax_checkers/python/vimpython3.vim
+++ b/syntax_checkers/python/vimpython3.vim
@@ -20,7 +20,7 @@ set cpo&vim

 function! SyntaxCheckers_python_vimpython3_IsAvailable() dict " {{{1
     if has('python3')
-        python import vim
+        python3 import vim
         return 1
     endif
     return 0
```

If I recall, the old pyflakes vim plugin I used to use was in memory. And the checks were done as I typed.

**Comment by lcd047** [comment]
> `vimpython3` seems to work for me. But the first time I run `:SyntasticInfo`, it gives the following error:

Thank you.  I suppose I didn't catch that because my Vim is compiled with suport for both Python 2 and Python 3.  Fixed in f3223b9.

> If I recall, the old pyflakes Vim plugin I used to use was in memory. And the checks were done as I typed.

Yes, this should work for all checkers that can be run as modules for one of the embedded interpreters.  And all python checkers can do that, at least in principle (well, I suppose some of them need a longer incantation than others).  Actually implementing the changes would be an interesting exercise, but so far there seems to be exactly zero interest for that, and YouCompleteMe already works fine.  My main motivation is Perl rather than Python anyway (and sadly Vim's support for Perl is way behind the support for Python).

**Comment by lcd047** [comment]
Ok, branch deleted.
